src/controller/CommissioningWindowOpener.cpp

```diff
@@ -124,6 +136,37 @@ CHIP_ERROR CommissioningWindowOpener::OpenCommissioningWindow(NodeId deviceId, S
     return mController->GetConnectedDevice(mNodeId, &mDeviceConnected, &mDeviceConnectionFailure);
 }
 
+CHIP_ERROR CommissioningWindowOpener::OpenCommissioningWindow(const CommissioningWindowVerifierParams & params,
+                                                              Callback::Callback<OnOpenCommissioningWindowWithVerifier> * callback)
+{
+    VerifyOrReturnError(mNextStep == Step::kAcceptCommissioningStart, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(kSpake2p_Min_PBKDF_Iterations <= params.GetIteration() &&
+                            params.GetIteration() <= kSpake2p_Max_PBKDF_Iterations,
+                        CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(params.GetSalt().size() >= kSpake2p_Min_PBKDF_Salt_Length &&
+                            params.GetSalt().size() <= kSpake2p_Max_PBKDF_Salt_Length,
+                        CHIP_ERROR_INVALID_ARGUMENT);
+    memcpy(mPBKDFSaltBuffer, params.GetSalt().data(), params.GetSalt().size());
+    mPBKDFSalt = ByteSpan(mPBKDFSaltBuffer, params.GetSalt().size());
+
+    ReturnErrorOnFailure(mVerifier.Deserialize(params.GetVerifier()));
+    mCommissioningWindowVerifierCallback = callback;
+    mBasicCommissioningWindowCallback    = nullptr;
+    mCommissioningWindowCallback         = nullptr;
+    mNodeId                              = params.GetNodeId();
+    mCommissioningWindowTimeout          = params.GetTimeout();
+    mPBKDFIterations                     = params.GetIteration();
+    mCommissioningWindowOption           = CommissioningWindowOption::kTokenWithProvidedPIN;
+    mSetupPayload                        = SetupPayload();
+    mSetupPayload.version                = 0;
+    mSetupPayload.discriminator.SetLongValue(params.GetDiscriminator());
+    mSetupPayload.rendezvousInformation.SetValue(RendezvousInformationFlag::kOnNetwork);
+
+    mNextStep = Step::kOpenCommissioningWindow;
+
+    return mController->GetConnectedDevice(mNodeId, &mDeviceConnected, &mDeviceConnectionFailure);
+}
+
 CHIP_ERROR CommissioningWindowOpener::OpenCommissioningWindowInternal(Messaging::ExchangeManager & exchangeMgr,
                                                                       const SessionHandle & sessionHandle)
```

Comment:
All operations are still targeted to RootEndpoint, how the bridge route OCW command to the bridged device after receiving it?

Reply:
In this context, there is no concept of directing at a specific endpoint once the bridge has mapped the original command (or any requests in general) from the received endpoint to the respective node being bridged. The aforementioned code is being used by controller part of fabric-admin app, which is forwarding a command to its connected device referenced by node id at the root endpoint.

Here's how it's working in the example fabric-admin and fabric-bridge apps (mostly for posterity and others reference since I realize you're already aware of this):


A. Fabric admin commissions a device with nodeId 2

```
[1720416217.588] [2568037:2568040] [CTL] Commissioning complete for node ID 0x0000000000000002: success
New device with Node ID: 0x2 has been successfully added.
[1720416217.588] [2568037:2568040] [-] AddSynchronizedDevice
[1720416217.658] [2568037:2568042] [-] AddSynchronizedDevice RPC call succeeded!
```

B. Fabric bridge adds the device to its parts list (maps nodeId 2 to endpoint 3)
```
[1720416217.632] [2568020:2568022] [-] Received AddSynchronizedDevice: 0000000000000002
[1720416217.658] [2568020:2568022] [-] Added device with nodeId=0x0000000000000002 to dynamic endpoint 3 (index=1)
```


C. Fabric bridge receives the OCW request directed at endpoint 3
```
[1720416217.745] [2568020:2568020] [DMG] InvokeRequestMessage =
[1720416217.745] [2568020:2568020] [DMG] {
[1720416217.745] [2568020:2568020] [DMG] 	suppressResponse = false,
[1720416217.745] [2568020:2568020] [DMG] 	timedRequest = true,
[1720416217.745] [2568020:2568020] [DMG] 	InvokeRequests =
[1720416217.745] [2568020:2568020] [DMG] 	[
[1720416217.745] [2568020:2568020] [DMG] 		CommandDataIB =
[1720416217.745] [2568020:2568020] [DMG] 		{
[1720416217.745] [2568020:2568020] [DMG] 			CommandPathIB =
[1720416217.745] [2568020:2568020] [DMG] 			{
[1720416217.745] [2568020:2568020] [DMG] 				EndpointId = 0x3,
[1720416217.745] [2568020:2568020] [DMG] 				ClusterId = 0x3c,
[1720416217.745] [2568020:2568020] [DMG] 				CommandId = 0x0,
[1720416217.745] [2568020:2568020] [DMG] 			},
[1720416217.745] [2568020:2568020] [DMG]
[1720416217.745] [2568020:2568020] [DMG] 			CommandFields =
[1720416217.745] [2568020:2568020] [DMG] 			{
[1720416217.745] [2568020:2568020] [DMG] 				0x0 = 900 (unsigned),
[1720416217.745] [2568020:2568020] [DMG] 				0x1 = [
[1720416217.745] [2568020:2568020] [DMG] 						0x73, 0xff, 0x12, 0xc4, 0x33, 0x84, 0x08, 0xf2, 0xff, 0xa0, 0x25, 0x0c, 0x15, 0x8a, 0xc4, 0x98, 0xd6, 0x9b, 0xb5, 0x56, 0x59, 0x7e, 0x25, 0x3c, 0x58, 0xc9, 0x87, 0xbc, 0x8c, 0x0b, 0x0c, 0xaf, 0x04, 0x09, 0x96, 0x35, 0x75, 0x2b, 0xac, 0xc1, 0x17, 0xaf, 0xfc, 0x81, 0xdc, 0x9e, 0x4f, 0xa4, 0x0b, 0xca, 0xc1, 0x64, 0xb7, 0xa4, 0xb0, 0x9e, 0xa5, 0x2b, 0xfc, 0x40, 0x44, 0x24, 0xc8, 0x09, 0x45, 0x56, 0x84, 0x76, 0xe1, 0x53, 0x35, 0x47, 0x6d, 0xb2, 0xb9, 0x36, 0x59, 0x4f, 0xd1, 0xbd, 0xad, 0x73, 0x29, 0x9f, 0xfd, 0xb7, 0x41, 0x74, 0x6e, 0x6a, 0x0a, 0x49, 0x9f, 0xf1, 0x5c, 0x3c, 0x97,
[1720416217.745] [2568020:2568020] [DMG] 				] (97 bytes)
[1720416217.745] [2568020:2568020] [DMG] 				0x2 = 995 (unsigned),
[1720416217.745] [2568020:2568020] [DMG] 				0x3 = 10000 (unsigned),
[1720416217.745] [2568020:2568020] [DMG] 				0x4 = [
[1720416217.745] [2568020:2568020] [DMG] 						0xd8, 0x07, 0x0d, 0xbf, 0x38, 0xaf, 0x85, 0xef, 0x1e, 0x3b, 0x30, 0x2d, 0xfd, 0x68, 0x77, 0x54, 0xd3, 0xcf, 0xe3, 0x05, 0x7c, 0x13, 0x09, 0x87, 0xa6, 0x0f, 0x51, 0xcb, 0x43, 0x9e, 0xaf, 0x62,
[1720416217.745] [2568020:2568020] [DMG] 				] (32 bytes)
[1720416217.745] [2568020:2568020] [DMG] 			},
[1720416217.745] [2568020:2568020] [DMG] 		},
[1720416217.745] [2568020:2568020] [DMG]
[1720416217.745] [2568020:2568020] [DMG] 	],
[1720416217.745] [2568020:2568020] [DMG]
[1720416217.745] [2568020:2568020] [DMG] 	InteractionModelRevision = 11
[1720416217.745] [2568020:2568020] [DMG] },
...
[1720416217.745] [2568020:2568020] [-] Received command to open commissioning window on Endpoint: 3
```

D. Fabric bridge looks up its mapping for bridged devices (endpointId 3 -> nodeId 2) and generates the RPC call to fabric-admin app for that node:
```
[1720416217.745] [2568020:2568020] [-] OpenCommissioningWindow with Node Id 0x:0000000000000002
```

E. Fabric admin receives the RPC call and pushes it as a command (2 is the node id for device on fabric-admin controller)
```
[1720416217.788] [2568037:2568038] [-] Received OpenCommissioningWindow request: 0x2
[1720416217.788] [2568037:2568038] [-] PushCommand: pairing open-commissioning-window 2 0 1 900 10000 995 --salt hex:d8070dbf38af85ef1e3b302dfd687754d3cfe3057c130987a60f51cb439eaf62 --verifier hex:73ff12c4338408f2ffa0250c158ac498d69bb556597e253c58c987bc8c0b0caf04099635752bacc117affc81dc9e4fa40bcac164b7a4b09ea52bfc404424c80945568476e15335476db2b936594fd1bdad73299ffdb741746e6a0a499ff15c3c97
[1720416217.788] [2568037:2568037] [-] Command: pairing open-commissioning-window 2 0 1 900 10000 995 --salt hex:d8070dbf38af85ef1e3b302dfd687754d3cfe3057c130987a60f51cb439eaf62 --verifier hex:73ff12c4338408f2ffa0250c158ac498d69bb556597e253c58c987bc8c0b0caf04099635752bacc117affc81dc9e4fa40bcac164b7a4b09ea52bfc404424c80945568476e15335476db2b936594fd1bdad73299ffdb741746e6a0a499ff15c3c97
```

F. Fabric Admin pushes the command to the respective device using the `CommissioningWindowOpener` in question (notice the endpoint is 0 as expected for the node itself):
```
[1720416217.792] [2568037:2568040] [EM] <<< [E:15423i S:63557 M:145490627 (Ack:195235134)] (S) Msg TX to 1:0000000000000002 [F6C8] [UDP:[fe80::f184:6119:bc23:f7f0%tailscale0]:5549] --- Type 0001:08 (IM:InvokeCommandRequest)
[1720416217.792] [2568037:2568040] [DMG] ICR moving to [AwaitingRe]
[1720416217.868] [2568037:2568040] [EM] >>> [E:15423i S:63557 M:195235135 (Ack:145490627)] (S) Msg RX from 1:0000000000000002 [F6C8] --- Type 0001:09 (IM:InvokeCommandResponse)
[1720416217.868] [2568037:2568040] [EM] Found matching exchange: 15423i, Delegate: 0x706a6c0182e8
[1720416217.868] [2568037:2568040] [EM] Rxd Ack; Removing MessageCounter:145490627 from Retrans Table on exchange 15423i
[1720416217.868] [2568037:2568040] [DMG] ICR moving to [ResponseRe]
[1720416217.868] [2568037:2568040] [DMG] InvokeResponseMessage =
[1720416217.868] [2568037:2568040] [DMG] {
[1720416217.869] [2568037:2568040] [DMG] 	suppressResponse = false,
[1720416217.869] [2568037:2568040] [DMG] 	InvokeResponseIBs =
[1720416217.869] [2568037:2568040] [DMG] 	[
[1720416217.869] [2568037:2568040] [DMG] 		InvokeResponseIB =
[1720416217.869] [2568037:2568040] [DMG] 		{
[1720416217.869] [2568037:2568040] [DMG] 			CommandStatusIB =
[1720416217.869] [2568037:2568040] [DMG] 			{
[1720416217.869] [2568037:2568040] [DMG] 				CommandPathIB =
[1720416217.869] [2568037:2568040] [DMG] 				{
[1720416217.869] [2568037:2568040] [DMG] 					EndpointId = 0x0,
[1720416217.869] [2568037:2568040] [DMG] 					ClusterId = 0x3c,
[1720416217.869] [2568037:2568040] [DMG] 					CommandId = 0x0,
[1720416217.869] [2568037:2568040] [DMG] 				},
[1720416217.869] [2568037:2568040] [DMG]
[1720416217.869] [2568037:2568040] [DMG] 				StatusIB =
[1720416217.869] [2568037:2568040] [DMG] 				{
[1720416217.869] [2568037:2568040] [DMG] 					status = 0x00 (SUCCESS),
[1720416217.869] [2568037:2568040] [DMG] 				},
[1720416217.869] [2568037:2568040] [DMG]
[1720416217.869] [2568037:2568040] [DMG] 			},
[1720416217.869] [2568037:2568040] [DMG]
[1720416217.869] [2568037:2568040] [DMG] 		},
[1720416217.869] [2568037:2568040] [DMG]
[1720416217.869] [2568037:2568040] [DMG] 	],
[1720416217.869] [2568037:2568040] [DMG]
[1720416217.869] [2568037:2568040] [DMG] 	InteractionModelRevision = 11
[1720416217.869] [2568037:2568040] [DMG] },
[1720416217.869] [2568037:2568040] [DMG] Received Command Response Status for Endpoint=0 Cluster=0x0000_003C Command=0x0000_0000 Status=0x0
[1720416217.869] [2568037:2568040] [CTL] Successfully opened pairing window on the device
```

G. The end device gets the command on root endpoint
```
[1720416217.805] [2568097:2568097] [EM] >>> [E:15423r S:42682 M:145490627 (Ack:195235134)] (S) Msg RX from 1:000000000001B669 [F6C8] --- Type 0001:08 (IM:InvokeCommandRequest)
[1720416217.805] [2568097:2568097] [EM] Found matching exchange: 15423r, Delegate: 0x5b1a4e4c7020
[1720416217.805] [2568097:2568097] [EM] Rxd Ack; Removing MessageCounter:195235134 from Retrans Table on exchange 15423r
[1720416217.805] [2568097:2568097] [DMG] Timed following action arrived at 0x0000000011AB93C5: handler 0x5b1a4e4c7020 exchange 15423r
[1720416217.805] [2568097:2568097] [DMG] Handing timed invoke to IM engine: handler 0x5b1a4e4c7020 exchange 15423r
[1720416217.805] [2568097:2568097] [DMG] InvokeRequestMessage =
[1720416217.805] [2568097:2568097] [DMG] {
[1720416217.805] [2568097:2568097] [DMG] 	suppressResponse = false,
[1720416217.805] [2568097:2568097] [DMG] 	timedRequest = true,
[1720416217.805] [2568097:2568097] [DMG] 	InvokeRequests =
[1720416217.805] [2568097:2568097] [DMG] 	[
[1720416217.805] [2568097:2568097] [DMG] 		CommandDataIB =
[1720416217.805] [2568097:2568097] [DMG] 		{
[1720416217.805] [2568097:2568097] [DMG] 			CommandPathIB =
[1720416217.805] [2568097:2568097] [DMG] 			{
[1720416217.805] [2568097:2568097] [DMG] 				EndpointId = 0x0,
[1720416217.805] [2568097:2568097] [DMG] 				ClusterId = 0x3c,
[1720416217.805] [2568097:2568097] [DMG] 				CommandId = 0x0,
[1720416217.805] [2568097:2568097] [DMG] 			},
[1720416217.805] [2568097:2568097] [DMG]
[1720416217.805] [2568097:2568097] [DMG] 			CommandFields =
[1720416217.805] [2568097:2568097] [DMG] 			{
[1720416217.805] [2568097:2568097] [DMG] 				0x0 = 900 (unsigned),
[1720416217.805] [2568097:2568097] [DMG] 				0x1 = [
[1720416217.805] [2568097:2568097] [DMG] 					0x73, 0xff, 0x12, 0xc4, 0x33, 0x84, 0x08, 0xf2, 0xff, 0xa0, 0x25, 0x0c, 0x15, 0x8a, 0xc4, 0x98, 0xd6, 0x9b, 0xb5, 0x56, 0x59, 0x7e, 0x25, 0x3c, 0x58, 0xc9, 0x87, 0xbc, 0x8c, 0x0b, 0x0c, 0xaf, 0x04, 0x09, 0x96, 0x35, 0x75, 0x2b, 0xac, 0xc1, 0x17, 0xaf, 0xfc, 0x81, 0xdc, 0x9e, 0x4f, 0xa4, 0x0b, 0xca, 0xc1, 0x64, 0xb7, 0xa4, 0xb0, 0x9e, 0xa5, 0x2b, 0xfc, 0x40, 0x44, 0x24, 0xc8, 0x09, 0x45, 0x56, 0x84, 0x76, 0xe1, 0x53, 0x35, 0x47, 0x6d, 0xb2, 0xb9, 0x36, 0x59, 0x4f, 0xd1, 0xbd, 0xad, 0x73, 0x29, 0x9f, 0xfd, 0xb7, 0x41, 0x74, 0x6e, 0x6a, 0x0a, 0x49, 0x9f, 0xf1, 0x5c, 0x3c, 0x97,
[1720416217.805] [2568097:2568097] [DMG] 				] (97 bytes)
[1720416217.805] [2568097:2568097] [DMG] 				0x2 = 995 (unsigned),
[1720416217.805] [2568097:2568097] [DMG] 				0x3 = 10000 (unsigned),
[1720416217.805] [2568097:2568097] [DMG] 				0x4 = [
[1720416217.805] [2568097:2568097] [DMG] 					0xd8, 0x07, 0x0d, 0xbf, 0x38, 0xaf, 0x85, 0xef, 0x1e, 0x3b, 0x30, 0x2d, 0xfd, 0x68, 0x77, 0x54, 0xd3, 0xcf, 0xe3, 0x05, 0x7c, 0x13, 0x09, 0x87, 0xa6, 0x0f, 0x51, 0xcb, 0x43, 0x9e, 0xaf, 0x62,
[1720416217.805] [2568097:2568097] [DMG] 				] (32 bytes)
[1720416217.805] [2568097:2568097] [DMG] 			},
[1720416217.805] [2568097:2568097] [DMG] 		},
[1720416217.805] [2568097:2568097] [DMG]
[1720416217.805] [2568097:2568097] [DMG] 	],
[1720416217.805] [2568097:2568097] [DMG]
[1720416217.805] [2568097:2568097] [DMG] 	InteractionModelRevision = 11
[1720416217.805] [2568097:2568097] [DMG] },
[1720416217.805] [2568097:2568097] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_003C e=0 p=a
[1720416217.805] [2568097:2568097] [DMG] AccessControl: allowed
[1720416217.805] [2568097:2568097] [DMG] Received command for Endpoint=0 Cluster=0x0000_003C Command=0x0000_0000
[1720416217.805] [2568097:2568097] [ZCL] Received command to open commissioning window
```

Complete logs:
[light-app.log](https://github.com/user-attachments/files/16146080/light-app.log)
[fabric-bridge.log](https://github.com/user-attachments/files/16146081/fabric-bridge.log)
[fabric-admin.log](https://github.com/user-attachments/files/16146082/fabric-admin.log)

Reply:
Yes, in your step C. Fabric bridge receives the OCW request directed at endpoint 3

I was not able to understand how 'OCW request directed at endpoint 3' is generated at the soruce Fabric Admin if we don't modify OpenCommissioningWindowInternal to specify the target endpoint

Reply:
Yeah, the `OpenCommissioningWindowInternal` isn't really invoked at that step by source fabric-admin.

Essentially, step C is getting the endpoint id [here](https://github.com/samadDotDev/connectedhomeip/blob/b154a1b651f627afd514d4a476201ae62d3bdf7f/examples/fabric-bridge-app/linux/main.cpp#L126), then for step D, fabric bridge looks up the device using endpoint id [here](https://github.com/samadDotDev/connectedhomeip/blob/b154a1b651f627afd514d4a476201ae62d3bdf7f/examples/fabric-bridge-app/linux/main.cpp#L147), which is then used to provide the node id to fabric admin over RPC as propagated from [here](https://github.com/samadDotDev/connectedhomeip/blob/b154a1b651f627afd514d4a476201ae62d3bdf7f/examples/fabric-bridge-app/linux/main.cpp#L152).

Eventually, at step F, fabric-admin calls `CommissioningWindowOpener::OpenCommissioningWindow()` [here](https://github.com/samadDotDev/connectedhomeip/blob/b154a1b651f627afd514d4a476201ae62d3bdf7f/examples/fabric-admin/commands/pairing/OpenCommissioningWindowCommand.cpp#L39) (which leads to invoking `OpenCommissioningWindowInternal` at the next step driven by the state machine there) and that point the endpoint 3 isn't important since that context only exists in the bridge, and the controller is already aware of the device (by its node id reference) on which this OCW command is being invoked on the device's root.

Let me know if you have any concerns or if there is anything I should be addressing as part of this PR (or in a follow-up one)?

Reply:
Discussed offline, we will continue to use https://github.com/project-chip/connectedhomeip/pull/33799 to track adding endpoint id parameter to the params. 

Scope to the current PR to the ECM params change.